src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1821,108 +1809,69 @@ private static ITypeSymbol GetMemberType(ISymbol memberSymbol)
-                    return;
-                }
-                else if (ancestor is AnonymousMethodExpressionSyntax delegateExpression)
-                {
-                    // If we are inside a delegate then use the return type of the Invoke Method of the delegate type
-                    var delegateType = InferTypes(delegateExpression).FirstOrDefault().InferredType;
-                    if (delegateType != null && delegateType.IsDelegateType())
-                    {
-                        var delegateInvokeMethod = delegateType.GetDelegateType(this.Compilation).DelegateInvokeMethod;
-                        if (delegateInvokeMethod != null)
-                        {
-                            types = SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(delegateInvokeMethod.ReturnType));
-                            isAsync = delegateExpression.AsyncKeyword.Kind() != SyntaxKind.None;
-                            return;
-                        }
-                    }
-                }
-                else if (ancestor is LocalFunctionStatementSyntax localFunctionStatement)
-                {
-                    // If we are inside a local function then use the return type of the local function
-                    var methodSymbol = (IMethodSymbol)SemanticModel.GetDeclaredSymbol(localFunctionStatement);
-                    types = SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(methodSymbol.ReturnType));
-                    isAsync = methodSymbol.IsAsync;
-                    return;
-                }
+                return ancestor is AnonymousFunctionExpressionSyntax anonymousFunction
+                    ? InferTypeInAnonymousFunctionExpression(anonymousFunction)
+                    : InferTypeInDeclaration(ancestor);
+            }
 
-                var memberSymbol = GetDeclaredMemberSymbolFromOriginalSemanticModel(SemanticModel, returnStatement.GetAncestorOrThis<MemberDeclarationSyntax>());
+            private IEnumerable<TypeInferenceInfo> InferTypeInArrowExpressionClause(ArrowExpressionClauseSyntax arrowClause)
+                => InferTypeInDeclaration(arrowClause.Parent);
 
-                switch (memberSymbol)
-                {
-                    case IMethodSymbol method:
-                        isAsync = method.IsAsync;
-                        types = SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(method.ReturnType));
-                        return;
-                    case IPropertySymbol property:
-                        types = SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(property.Type));
-                        return;
-                    case IFieldSymbol field:
-                        types = SpecializedCollections.SingletonEnumerable(new TypeInferenceInfo(field.Type));
-                        return;
-                }
+            private IEnumerable<TypeInferenceInfo> InferTypeInDeclaration(SyntaxNode declaration)
```

Comment:
this is pretty vague.  can we have a comment specifying what the expected declaration types would be here?

Reply:
This should work for pretty much anything that returns a GetDeclaredSymbol. It *has to* work for (meaning there are tests) all base-method members, accessors & local functions. Should I add a comment enumerating this?

Reply:
Actually, forget my previous comment... I thought it was about `GetDeclaredMemberSymbolFromOriginalSemanticModel`

Reply:
➡️ Fixed in 220ce00b3a9e4098e2bb0f3285a3ca8196fd7dfd